[Doc] API documentation updates.

dan-zheng opened the pull request:
Assorted documentation updates.

dan-zheng committed:
[Doc] API documentation updates.

mhong commented on stdlib/public/Python/NumpyConversion.swift:
```diff
@@ -93,6 +93,13 @@ extension Double : NumpyScalarCompatible {
+  ///
+  /// - Parameter numpyArray: The `numpy.ndarray` instance to convert.
+  /// - Precondition: The `numpy` Python package must be installed.
+  /// - Precondition: `numpyArray` must be 1-D and have a compatible scalar
```
do you want to more explicitly describe when a nil value will be returned? Same for other APIs below.

`Precondition` usually describes cases where the code will lead to undefined behavior (e.g. crash) when the condition is violated.

rxwei replied:
+1. Precondition is definitely not the right term for describing `nil`. We should look at how common failable initializers are documented in the standard library.

dan-zheng replied:
Thanks for great catch!

A `Returns:` comment is typically used to describe failable initializers and when they return `nil`.

Updated in https://github.com/apple/swift/pull/23069/commits/e7920e443eeb844ac662c91069d9c9025a3c2439, following [this stdlib documentation](https://github.com/apple/swift/blob/948655e850e0a4dd8c0f6d5a86f20f2c462c7303/stdlib/public/core/CString.swift#L137).

mhong reviewed (approved):
nice!

rxwei commented on stdlib/public/TensorFlow/NumpyConversion.swift:
```diff
@@ -87,6 +93,12 @@ extension ShapedArray : ConvertibleFromNumpyArray
+  ///
+  /// - Parameter numpyArray: The `numpy.ndarray` instance to convert.
+  /// - Precondition: The `numpy` Python package must be installed.
+  /// - Precondition: `numpyArray` must have a compatible scalar datatype.
```
"datatype" isn't a thing in either Numpy or Swift. `dtype` (backquoted) is probably more appropriate.

```suggestion
  /// - Precondition: `numpyArray` must have a compatible scalar datatype.
```

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/23069/commits/e7920e443eeb844ac662c91069d9c9025a3c2439.

rxwei commented on stdlib/public/Python/NumpyConversion.swift:
```diff
@@ -93,6 +93,13 @@ extension Double : NumpyScalarCompatible {
+  /// - Parameter numpyArray: The `numpy.ndarray` instance to convert.
+  /// - Precondition: The `numpy` Python package must be installed.
+  /// - Precondition: `numpyArray` must be 1-D and have a compatible scalar
+  ///   datatype.
```
```suggestion
  ///   `dtype`.
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/23069/commits/e7920e443eeb844ac662c91069d9c9025a3c2439.

dan-zheng committed:
Address feedback.

- Change precondition comment to return value comment for failable initializers.
- Use `dtype` instead of "datatype".

Address feedback from @mhong and @rxwei.

dan-zheng commented:
@swift-ci Please test tensorflow
